### DISCUSSION TITLE
Garbage Collection specifics.

### DISCUSSION THREAD

**Discussion by sasq64** [mruby/mruby#6234] [Q&A] [answered]
I am trying to understand how the GC works in more detail, since I want to make sure the lifetime of objects shared between Ruby and C works as I expect. Is this roughly correct;

* GC preallocates heaps with pointers to RValues, and creates free lists.
* When objects are created, a new pointer is taken from the free list.
* When GC is performed, first all live objects are marked by traversing tree from roots.
* Then all objects in all heaps are iterated, and objects that are not alive are freed.
* In addition, there is a GC arena used to temporarily store objects created in C (`mrb_obj_alloc()`). Objects in the arena are not collected.
* The arena is cleared when returning to Mruby VM after call into registered C function.

Regarding the Arena, would it have been possible to instead just turn off GC when inside C functions, and turn it back on when returning?

**Comment by matz** [comment, marked as answer]
Your basic understanding of GC is correct, except that does not cover the generational and incremental aspects of the garbage collector.

Maybe `doc/guides/gc-arena-howto.md` would help you to understand the arena better.

And the answer to your question would be "yes, but it may be hard to detect allocating too many objects".
